### PR TITLE
fix: add prop to render to skip configureHostComponentNamesIfNeeded

### PR DESCRIPTION
**Pull request by KrastanD**
### Summary

https://github.com/callstack/react-native-testing-library/issues/1423

### Test plan

This change can be tested by running the renderHook.test.tsx tests after clearing the jest cache. Before my change, the first test takes 445 ms and after my change it takes 5 ms. 

Before: 
``` ~/Code/react-native-testing-library   main  yarn jest ./src/__tests__/renderHook.test.tsx
yarn run v1.22.19
$ ~/Code/react-native-testing-library/node_modules/.bin/jest ./src/__tests__/renderHook.test.tsx
 PASS  src/__tests__/renderHook.test.tsx
  ✓ gives comitted result (445 ms)
  ✓ allows rerendering (1 ms)
  ✓ allows wrapper components (2 ms)

Test Suites: 1 passed, 1 total
Tests:       3 passed, 3 total
Snapshots:   0 total
Time:        1.273 s
Ran all test suites matching /.\/src\/__tests__\/renderHook.test.tsx/i.
✨  Done in 2.19s.
```

After: 
```~/Code/react-native-testing-library   fix/renderHook-skip-component-names-config  yarn jest ./src/__tests__/renderHook.test.tsx 
yarn run v1.22.19
$ ~/Code/react-native-testing-library/node_modules/.bin/jest ./src/__tests__/renderHook.test.tsx
 PASS  src/__tests__/renderHook.test.tsx
  ✓ gives comitted result (5 ms)
  ✓ allows rerendering (1 ms)
  ✓ allows wrapper components

Test Suites: 1 passed, 1 total
Tests:       3 passed, 3 total
Snapshots:   0 total
Time:        0.819 s
Ran all test suites matching /.\/src\/__tests__\/renderHook.test.tsx/i.
✨  Done in 1.74s.
```

### CODE REVIEW

**Comment by pierrezimmermannbam** [comment]
Thanks for opening this pr @KrastanD!  Regarding the implementation, I don't think the API of `render` should change because it won't work if the host names detection is not run. Also it never needs to be executed in `renderHook` so it does not look like we need this to be configurable since in one case we always need it and in the other one we never do. Could you update your pr to keep the public API and just make it so that `renderHook` doesn't run name detection?

This would be a good step forward but it would only make `renderHook` faster though so it doesn't fully solve the problem. I wonder if we should allow host component names to be configured by the user. It would give more control to users. On the other hand I'm not sure it should be part of the public API. I'd probably use it but it also makes the API more complex so I'm not sure it's worth it. Wdyt @mdjastrzebski @MattAgn  @AugustinLF ?

**Comment by KrastanD** [comment]
Hopefully this update is more of what you had in mind @pierrezimmermannbam. I pulled out only the necessary parts of the render function into the renderHook. 

I didn't want to change the api in any way so there is a bit of code duplication between `updateWithAct` in the render function and `baseRerender` in renderHook. With your blessing, I can pull out `updateWithAct` into its own file like `renderWithAct` is and reuse it in renderHook.

File: src/renderHook.tsx [hunk 1]
```diff
@@ -42,12 +43,20 @@ export function renderHook<Result, Props>(
     return null;
   }
 
-  const { rerender: baseRerender, unmount } = render(
+  const wrap = (element: React.ReactElement) =>
```
Comment (mdjastrzebski) [inline src/renderHook.tsx:46]: What is the purpose of this change? It seems to do the same thing as previous code with just passed wrapper to `renderWithAct`

**Comment by mdjastrzebski** [comment]
@KrastanD I support resolving the issue of `renderHook` unnecessarily calling `detectHostComponents`. I've encountered another use-case when it would be beneficial: in case user mocks `react-native` package using `jest.mock()` and then tests hook, the `renderHook` fails as `Text`/`TestInput` is not longer available (user just mocked whole `react-native` package), but the hook itself does not need it at all. 

Current implementation feels unnecessarily complex though. I would opt for "internal" (non-documented) `render` flag that would control `detectHostComponents` method being called. By default it would be called (when user just invokes `render`), but when `renderHook` would call `render` it would use it to ignore that step.

Proposed naming:
* `render` option name: `detectHostComponentNames`

@pierrezimmermannbam, @KrastanD wdyt?

**Comment by KrastanD** [comment]
@mdjastrzebski My initial version of this PR which you can see [here](https://github.com/callstack/react-native-testing-library/pull/1425/commits/ccd01ffc616d580fb35eded1c29bff70eaa59e8a) solved the problem in a similar way to what you are suggesting except instead of naming it `detectHostComponentNames` i named it `skipHostComponentNamesConfiguration` because by default it would be set to false. In @pierrezimmermannbam's comment above, he says the API of `render` shouldn't change which is why I rewrote it to use the internals of `render`, like `renderWithAct`, but without the hostComponentNames detection.

File: src/renderHook.tsx (hunk 1 above)
Comment (KrastanD) [reply src/renderHook.tsx:46]: `renderWithAct` does what `render` does but without detecting hostComponentNames

**Comment by pierrezimmermannbam** [comment]
@mdjastrzebski  @KrastanD I personally like this implementation, I don't think it's complex, in fact I think it's an improvement. Sure there is a bit of code duplication to handle wrapper option and update with act but it doesn't add an additional prop to the render function just for renderHook's sake. This is a better separation of concern imo, if tomorrow we're to add a new feature that's only for components we'll need to add one more flag. And with this implementation it's very easy to change renderHook or render without changing the behavior of the other one, which is what we want I believe. I think this implementation will be easier to maintain so I would keep it. Also even of the flag is not documented it would be visible to users with autocomplete, it's not a big problem but still a downside

**Review by mdjastrzebski — APPROVED** [review]

**Comment by mdjastrzebski** [comment]
@KrastanD, @pierrezimmermannbam I've refactored the code so that it uses internal option `detectHostComponentNames` which is available internally on `renderInternal` function but not exposed publicly. I've also added tests that ensure proper behavior in that regard for both `render` and `renderHook` public methods.

**Comment by mdjastrzebski** [comment]
This PR has been released in v12.1.3 🚢
